cdap-gateway/src/main/java/co/cask/cdap/gateway/discovery/UserServiceEndpointStrategy.java

```diff
@@ -39,48 +39,71 @@
   private final ProgramId serviceId;
   private final String version;
   private final Random random;
+  private final RouteFallbackStrategy fallbackStrategy;
 
   public UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId,
-                                     @Nullable String version) {
+                                     String fallbackStrategy, @Nullable String version) {
     super(serviceDiscovered);
     this.routeStore = routeStore;
     this.serviceId = serviceId;
     this.version = version;
     this.random = ThreadLocalRandom.current();
+    this.fallbackStrategy = RouteFallbackStrategy.valueOfRouteFallbackStrategy(fallbackStrategy);
   }
 
-  public UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId) {
-    this(serviceDiscovered, routeStore, serviceId, null);
+  @VisibleForTesting
+  UserServiceEndpointStrategy(ServiceDiscovered serviceDiscovered, RouteStore routeStore, ProgramId serviceId) {
+    this(serviceDiscovered, routeStore, serviceId, null, null);
```

Comment:
Instead of passing in `null` for fallback strategy, better pass in the default (which in RANDOM).